controllers/openshift_controller.go

```diff
@@ -1210,12 +1221,151 @@ func (eh *McpEventHandler) Generic(event event.GenericEvent, queue workqueue.Rat
 }
 
 
+func (r *KataConfigOpenShiftReconciler) nodeMatchesKataSelector(nodeLabels map[string]string) bool {
+	nodeSelector, err := r.getNodeSelectorAsSelector()
+
+	if err != nil {
+		r.Log.Info("couldn't get kata node selector", "err", err)
+		// If we cannot find out whether a Node matches assuming that it
+		// doesn't seems to be the safer assumption.  This also seems
+		// consistent with error-handling sematics of earlier similar code.
+		return false
+	}
+
+	return nodeSelector.Matches(labels.Set(nodeLabels))
+}
+
+func isWorkerNode(node client.Object) bool {
+	if _, ok := node.GetLabels()["node-role.kubernetes.io/worker"]; ok {
+		return true
+	}
+	return false
+}
+
+func getStringMapDiff(oldMap, newMap map[string]string) (added, modified, removed map[string]string) {
+	added = map[string]string{}
+	modified = map[string]string{}
+	removed = map[string]string{}
+
+	if oldMap == nil {
+		added = newMap
+		return
+	}
+
+	if newMap == nil {
+		removed = oldMap
+		return
+	}
+
+	for newKey, newVal := range newMap {
+		if _, ok := oldMap[newKey]; !ok {
+			added[newKey] = newVal
+		}
+	}
+	for oldKey, oldVal := range oldMap {
+		if _, ok := newMap[oldKey]; !ok {
+			removed[oldKey] = oldVal
+		} else {
+			if oldMap[oldKey] != newMap[oldKey] {
+				modified[oldKey] = newMap[oldKey]
+			}
+		}
+	}
+	return added, modified, removed
+}
+
+
+type NodeEventHandler struct {
+	reconciler *KataConfigOpenShiftReconciler
+}
+
+func (eh *NodeEventHandler) Create(event event.CreateEvent, queue workqueue.RateLimitingInterface) {
+	node := event.Object
+
+	log := eh.reconciler.Log.WithName("NodeCreate").WithValues("node name", node.GetName())
+	log.Info("node created")
+
+	if !isWorkerNode(node) {
```

Comment:
Actually, no. It's a bit tricky. This is how a converged cluster looks like:
```
oc get nodes
NAME                                        STATUS   ROLES                         AGE   VERSION
converged-412-ctlplane-0.converged412.com   Ready    control-plane,master,worker   45h   v1.25.4+18eadca
converged-412-ctlplane-1.converged412.com   Ready    control-plane,master,worker   45h   v1.25.4+18eadca
converged-412-ctlplane-2.converged412.com   Ready    control-plane,master,worker   45h   v1.25.4+18eadca

```

Note how all masters are also workers.  So this code still works on converged clusters although much of it is not really need in that environment.  Since `kataConfigPoolSelector` cannot be supported on converged clusters, the only operations possible are basically just install kata on all nodes and uninstall kata from all nodes.  The only useful part of node watching on converged clusters is triggering reconciliation on `machineconfiguration.openshift.io/state` changes, and that works.

Reply:
Actually you're right, I misread the condition in your comment, I apologize. We do want to avoid masters in normal clusters but not in converged ones, and we use the fact that all masters in converged clusters also have the "worker" node role label to implement that.